### PR TITLE
Added an API for a logged in user to retrieve metadata from database

### PR DESCRIPTION
**Pull request by imjalpreet**
@oparoz I have added the API for a logged in user. 

Can you check if its correct?

### CODE REVIEW

File: lib/Controller/RetrieveMetadataApiController.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+use OCA\MediaMetadata\Services\RetrieveMetadata;
+use OCP\AppFramework\ApiController;
+
+/**
+ * Class RetrieveMetadataApiController
+ *
+ * @package OCA\MediaMetadata\Controller
+ */
+class RetrieveMetadataApiController extends ApiController {
```
Comment (oparoz) [inline lib/Controller/RetrieveMetadataApiController.php]: For a logged-in user, you need to extend Controller

File: appinfo/routes.php [hunk 2]
```diff
@@ -19,7 +19,16 @@
  */
 return [
     'routes' => [
-	   ['name' => 'page#index', 'url' => '/', 'verb' => 'GET'],
-	   ['name' => 'page#do_echo', 'url' => '/echo', 'verb' => 'POST'],
+		['name' => 'page#index', 'url' => '/', 'verb' => 'GET'],
+		['name' => 'page#do_echo', 'url' => '/echo', 'verb' => 'POST'],
```
Comment (oparoz) [inline appinfo/routes.php]: Please remove those routes as your app is "headless" for now

File: lib/Controller/RetrieveMetadataApiController.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	/**
+	 * @var RetrieveMetadata
+	 */
+	protected $retrievalService;
+
+	/**
+	 * @param RetrieveMetadata $retrieveMetadata
+	 */
+	public function __construct(RetrieveMetadata $retrieveMetadata) {
+		$this->retrievalService = $retrieveMetadata;
```
Comment (oparoz) [inline lib/Controller/RetrieveMetadataApiController.php]: You're missing a call to the parent method `parent::__construct($appName, $request);`

File: lib/Controller/RetrieveMetadataApiController.php [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+	/**
+	 * @var RetrieveMetadata
+	 */
+	protected $retrievalService;
+
+	/**
+	 * @param RetrieveMetadata $retrieveMetadata
+	 */
+	public function __construct(RetrieveMetadata $retrieveMetadata) {
```
Comment (oparoz) [inline lib/Controller/RetrieveMetadataApiController.php]: I think for automatic injection to work, you'll need to use $Appname with a capital A.

File: lib/Controller/RetrieveMetadataApiController.php (hunk 1 above)
Comment (oparoz) [inline lib/Controller/RetrieveMetadataApiController.php]: And the name needs to be MetadataController if you want to match the route.

https://docs.nextcloud.com/server/10/developer_manual/app/routes.html

**Comment by imjalpreet** [comment]
@oparoz Changes Done.

File: lib/Controller/MetadataController.php [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+		$this->retrievalService = $retrieveMetadata;
+	}
+
+	/**
+	 * @NoAdminRequired
+	 *
+	 * @param $fileList
+	 * @return array
+	 */
+	public function getMetadata($fileList) {
```
Comment (oparoz) [inline lib/Controller/MetadataController.php]: How are you going to send the filelist to the controller? You can't send PHP objects. The list is going to be sent via GET, so you need to process what you get and send it to the service in a format it understands (Array)

Comment (imjalpreet) [inline lib/Controller/MetadataController.php]: @oparoz Sorry, I forgot to do this. But I have added that now.

**Comment by oparoz** [comment]
You need to add the unit test to make sure things are working as expected.

In the next PR, you'll need to enable the API suite in Codeception to make sure the app is working as expected. Everything you need is in the Gallery app.
You will need a shorter version of this to manage users and files:
https://github.com/interfasys/galleryplus/blob/master/tests/_support/Helper/DataSetup.php

**Comment by imjalpreet** [comment]
@oparoz What should I use in $AppName when I have to test?

**Comment by oparoz** [comment]
Any string

**Comment by imjalpreet** [comment]
@oparoz I will again have to use setMapperResult, right?

If yes, then I have tried it but I am getting the following error: 

```

1) MetadataControllerTest: Get metadata
 Test  tests/unit/Controller/MetadataControllerTest.php:testGetMetadata
Array (...) does not match expected type "NULL".
#1  /var/www/GSoC/apps/mediametadata/tests/unit/Controller/MetadataControllerTest.php:80
#2  OCA\MediaMetadata\Controller\MetadataControllerTest->testGetMetadata

```

**Comment by oparoz** [comment]
> I will again have to use setMapperResult, right?

Not this time. You simply need to mock the service. It's much easier since it returns an array.

**Comment by imjalpreet** [comment]
@oparoz Done.

File: tests/unit/Controller/MetadataControllerTest.php [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+ *
+ * @author Jalpreet Singh Nanda (:imjalpreet) <jalpreetnanda@gmail.com>
+ * @copyright Jalpreet Singh Nanda (:imjalpreet) 2016
+ */
+
+namespace OCA\MediaMetadata\Controller;
+
+use \OCA\MediaMetadata\Tests\MapperTestUtility;
+
+class MetadataControllerTest extends MapperTestUtility {
```
Comment (oparoz) [inline tests/unit/Controller/MetadataControllerTest.php]: You don't need to extend this class

File: tests/unit/Controller/MetadataControllerTest.php [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+	public function setUp() {
+		parent::setUp();
+
+		$this->request = $this->getMockBuilder('\OCP\IRequest')
+			->disableOriginalConstructor()
+			->getMock();
+
+		$this->retrieveMetadata = $this->getMockBuilder('\OCA\MediaMetadata\Services\RetrieveMetadata')
+			->setConstructorArgs([
+				$this->db
```
Comment (oparoz) [inline tests/unit/Controller/MetadataControllerTest.php]: You need to disable the constuctor here like for the other Classes

Comment (imjalpreet) [inline tests/unit/Controller/MetadataControllerTest.php]: @oparoz Why should this be done?

**Comment by imjalpreet** [comment]
@oparoz Done.

**Comment by oparoz** [comment]
CI didn't kick in. Could you please force push the last commit again?

**Comment by oparoz** [comment]
I'm merging this as-is because we're pressed for time and you need to enable the API suite, but try to think about what your service should do in case of error.

You could get:
- text instead of a fileId
- a fileId which doesn't exist
- a fileID which doesn't have any metadata (a PDF per example)

All these will probably crash your app and we need to make it return an error message instead.

**Comment by imjalpreet** [comment]
@oparoz Yeah, I understand. I will try to think of all these errors and maybe make a list, so that I can add them afterwards.

**Comment by oparoz** [comment]
Thank you

**Pull request merged**